### ISSUE TITLE
Adding Table Transformer models to HuggingFace Transformers

### ISSUE THREAD

**Issue by NielsRogge** [microsoft/table-transformer#68]
Hi Table Transformer team :)

As I've implemented DETR in [🤗 HuggingFace Transformers](https://github.com/huggingface/transformers.git) a few months ago, it was relatively straightforward to port the 2 checkpoints you released. Here's a notebook that illustrates inference with DETR for table detection and table structure recognition: https://colab.research.google.com/drive/1lLRyBr7WraGdUJm-urUm_utArw6SkoCJ?usp=sharing

As you may or may not know, any model on the HuggingFace [hub](https://huggingface.co/) has its own Github repository. E.g. the DETR-table-detection checkpoint can be found here: https://huggingface.co/nielsr/detr-table-detection. If you check the "files and versions" tab, it includes the weights. The model hub uses git-LFS (large file storage) to use Git with large files such as model weights. This means that any model has its own Git [commit history](https://huggingface.co/nielsr/yolos-s/commits/main)!

A model card can also be added to the repo, which is just a README.

Are you interested in joining the Microsoft [organization](https://huggingface.co/microsoft) on the hub, such that we can store all model checkpoints there (rather than under my user name)?

Also, it would be great to add PubTables-1M (and potentially other datasets, useful for improving AI on unstructured documents) to the 🤗 hub. Would you be up for that?

Let me know!

Kind regards,

Niels
ML Engineer @ HuggingFace

**Comment by NielsRogge** [comment]
Relevant to #17, #44, #57

**Comment by NielsRogge** [comment]
cc'ing @bsmock for visibility

**Comment by light42** [comment]
@NielsRogge could you add another pipeline to turn the results into DataFrame?

**Comment by NielsRogge** [comment]
Hi,

The model isn't extracting any text from images. It's just an object detector, so it can detect where tables, table rows and columns etc. are in an image. You would need to crop the image based on the bounding box coordinates, after which you can run an OCR model (like [TrOCR](https://huggingface.co/docs/transformers/model_doc/trocr)) to get the actual text.

**Comment by salman-moh** [comment]
@NielsRogge is TrOCR better than tesseract? I thought pytesseract would be the top option

**Comment by bsmock** [comment]
Hi Niels,

We're excited to see that you were able to successfully integrate the model into HuggingFace Transformers. Nice work--this is great! We will review what you've done to make sure it's all correct and see how we can get involved. Thanks!

Cheers,
Brandon

**Comment by NielsRogge** [comment]
> Is TrOCR better than tesseract? I thought pytesseract would be the top option

TrOCR seems better than Tesseract, especially on handwritten text. However, the model does only work on single text-line images (unlike Tesseract which works on entire PDFs).

> Also is TSR part of your co-lab notebook from the MSFT team table-transformer or just the TD section?
The Table Structure Recognition part is also from Microsoft (this repository shared 2 checkpoints; one for table detection and one for table structure recognition)

**Comment by salman-moh** [comment]
Any thoughts on https://github.com/microsoft/table-transformer/issues/69#issue-1368131851 @NielsRogge ?

**Comment by khadkechetan** [comment]
@NielsRogge This is awesome. However I guess we have to work on getting data in csv. Let me know for any reference for the same. 
Since this is the MIT license, we can use for commercial. Please confirm. 
 
![image](https://user-images.githubusercontent.com/38162582/189598192-4988bef9-1e2d-42db-904f-23e3a8fe55ff.png)

**Comment by vallabh001** [comment]
Hey @NielsRogge, @bsmock, can you please help me with this ..
#70

**Comment by bsmock** [comment]
@NielsRogge, I joined the Microsoft organization. Do we need to transfer any assets there, or how do you normally organize the assets between your profile and the Microsoft org?

**Comment by NielsRogge** [comment]
Hi @bsmock,

Thanks for joining the Microsoft organization. I've opened a PR to add Table Transformer here: #18920.

Once reviewed, I'll transfer the checkpoints currently hosted under my name to microsoft. You'll have write access, meaning that you can add model cards, update repos etc.

**Comment by emigomez** [comment]
Hi @NielsRogge 

One question regarding your notebook. How can I obtain the rows or each element bounding boxes?

In the table structure recognition you have:
```
postprocessed_outputs = feature_extractor.post_process(outputs, target_sizes)
    bboxes_scaled = postprocessed_outputs[0]['boxes'][keep]
```

And with that, I can only obtain the columns, but I would like to obtain all the other objects.
I suppose that this information is also returned by the network but I don't know how to obtain it.

Thanks in advance!

**Comment by yellowjs0304** [comment]
Hi, @NielsRogge Thank you for sharing nice work.
I tried your [work](https://colab.research.google.com/drive/1lLRyBr7WraGdUJm-urUm_utArw6SkoCJ?usp=sharing). 
But It returned different result compared with your works. I think the model loading is failed in some layers. 
![image](https://user-images.githubusercontent.com/30281608/194267655-7b9af272-8350-4be6-b3e5-689a17521a3d.png)

I don't know detailed reason, but maybe there's problem with transformer installation.  
I didn't install your private transformer (branch add_table_transformer). 
I just installed official transformer(v. 4.22.0.dev0 -> from pip install transformers).
Do i need to install by yours? not official?

Of course, the results is bad.
![image](https://user-images.githubusercontent.com/30281608/194267984-dac36dd1-8035-4b44-b49a-eb2f3d160395.png)

Could you please share any opinion about this?

**Comment by salman-moh** [comment]
Hi @yellowjs0304  I dont see why there is such a prediction when I just ran the entire notebook of Neils as is, and it works:
![image](https://user-images.githubusercontent.com/31631107/194279627-71516c55-6b73-44f3-9178-65de31126dd5.png)
Please check his code again

**Comment by yellowjs0304** [comment]
@salman-moh 
Hi, Thank you for sharing opinion.
Could you please tell me how did you installed transformers? Did you used the command lines in jupyter notebook?

**Comment by salman-moh** [comment]
yes, exactly how Neils did. Btw, I also have created a HuggingFace Space if you want to take a look :) https://huggingface.co/spaces/SalML/TableTransformer2CSV
What it does is, paste an image with a table, it will detect and give you a dataframe that can be downloaded as a CSV.
and yes its also based on the 4.22dev version. Do not install by `pip install transformers`

**Comment by yellowjs0304** [comment]
@salman-moh Hi, Thank you for fast reply... but, the problem is not solved yet..
I deleted and make new conda env, and followed Niels code too. 
![image](https://user-images.githubusercontent.com/30281608/194448359-a0ff8a30-4326-4fd0-a96c-e79a15dc83c5.png)

but the problem is still remain.. Could you please try it with new environment and same notebook file? 
I'm not familiar with notebook, So maybe it's not correct idea. but, I think the command `pip install -q ./transformers` it might be related with [official transformers](https://github.com/huggingface/transformers/commits/main/src/transformers/models/detr).

**Comment by emigomez** [comment]
Hi! and thank you for sharing nice work...
I think the output of the TSR model does not fill in the individual cells of the table. I suppose that the solution to find these cells is to obtain them from rows and columns bounding boxes, someone has carried out this procedure?

**Comment by salman-moh** [comment]
yes, that is what I've done underneath the [HF space](https://huggingface.co/spaces/SalML/TableTransformer2CSV), the code can be found here: https://github.com/salman-moh/TableTransformer2CSV/blob/salman-moh-patch-1/app.py or even from the [HF repo](https://huggingface.co/spaces/SalML/TableTransformer2CSV/tree/main)

**Comment by emigomez** [comment]
Thank you @salman-moh, to share your work with us!!

Another question is whether it is possible to just launch the TSR without any TD (the input image to the TSR would be the table extracted). I try to run just the nielsr/detr-table-structure-recognition model, but the thing is that always the output has a table detected and is usually smaller than the original one, is it possible to run nielsr/detr-table-structure-recognition just to detect structure from an input table image?

![ex](https://user-images.githubusercontent.com/64838403/195073861-127a10cd-5656-4d63-8ccf-c361ac85f3fc.PNG)

This is an example where the input is directly the table, and as you can see the nielsr/detr-table-structure-recognition obtains a table to perform the TSR.

**Comment by salman-moh** [comment]
This is way offtopic for the issue. Please see some of the issues related to padding, in short, pad the entire 4 edges of the table and that might help.

**Comment by emigomez** [comment]
@salman-moh thank you for your response, I know that including padding can help with the table detection issues, but the question is if it is possible to just obtain the table structure of an input table image, without any detection. I mean, if there is a parameter or something to run nielsr/detr-table-structure-recognition to obtain a result of structure over the full image, like in the next example:
![19](https://user-images.githubusercontent.com/64838403/195094191-9dda79b9-6909-4122-a502-a9e940739e43.jpg)

I don't know if this is the best place to put the issue (sorry for that), but I write it here because I think that it refers more to the notebook than to the current repo

**Comment by MartinHaus1993** [comment]
Thanks @NielsRogge, saw that you have just added this now on the repo. Can confirm this works now with transformers 4.24.0.dev0 and the new TableTransformerForObjectDetection.


Thanks!

**Comment by NielsRogge** [comment]
Also @bsmock would be great to add a mention in the README of this repo, as people probably will find it helpful ;)

**Comment by bsmock** [comment]
@NielsRogge, finally getting a chance to circle back to this. First thing I'm wondering is what is the best way to put PubTables-1M onto the hub? Would you recommend just adding tar.gz files like we have currently? Or can the HuggingFace hub support such a large object detection dataset in a more database-like way? Any pointers/documentation you have on uploading large image datasets would be great!

**Comment by pathikg** [comment]
> Hi @NielsRogge, is it possible to use the microsoft table-transformer pre-trained model to train it on my tables for a specific task? If yes, how to do it? Because I'm lost in the different documentations on HuggingFace or Github Thanks in advance

I think this tutorial might help
https://huggingface.co/docs/transformers/tasks/object_detection

- prepare your dataset similar to CPPE-5 dataset as shown in the tutorial
- replace checkpoint name with `microsoft/table-transformer-detection` since table transformer is identical to DETR
- use everything as it is

**Comment by aditya-ihx** [comment]
> @NielsRogge, finally getting a chance to circle back to this. First thing I'm wondering is what is the best way to put PubTables-1M onto the hub? Would you recommend just adding tar.gz files like we have currently? Or can the HuggingFace hub support such a large object detection dataset in a more database-like way? Any pointers/documentation you have on uploading large image datasets would be great!

Have you found a solution to this problem ? I am facing the same right now

**Comment by NielsRogge** [comment]
Hi folks, new checkpoints (with updated notebooks and demos) are now available at #158

**Comment by mohammadreza-sheykhmousa** [comment]
Hi @NielsRogge thanks for your work. To fine-tune the model on table detection I was wondering if I need to have the annotation in coco format? I appreciate if also let me know about the structure (in terms of directory) of input data -should it be 
-data
    ----training
        ------images
        ------annotations
    ----val
        ------images
        ------annotations
    ----testing
        ------images
        ------annotations

**Comment by NielsRogge** [comment]
Hi,

Yes if you follow my [fine-tuning notebook](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/DETR/Fine_tuning_DetrForObjectDetection_on_custom_dataset_(balloon).ipynb), your data needs to be in the COCO format.

Personally I'd recommend a tool like [RoboFlow](https://roboflow.com/) which allows to annotate data and export it into any format you prefer, like COCO.

**Comment by SohamTolwala** [comment]
Hi @NielsRogge I was using the [inference notebook](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/Table%20Transformer/Inference_with_Table_Transformer_(TATR)_for_parsing_tables.ipynb) just till the part of table detection and cropping the table out. The issue I had was that the model was detecting false tables too, here is an example of what it detected (a complete blank area). It is detecting the required tables correctly although this is an issue I want to tackle, would fine tuning resolve this? Please do suggest on this, maybe any parameters that I can tweak?
![table_0_page_3](https://github.com/microsoft/table-transformer/assets/94548689/8b5d4cd9-eb47-41fe-a791-097776e2be5b)
